src/server/scripts/Pandaria/zone_the_wandering_isle.cpp

```diff
@@ -0,0 +1,2356 @@
+
+        void EnterCombat(Unit* /*who*/) override
+        {
+            _events.ScheduleEvent(EVENT_THROW_ROCK, 0);
+        }
+
+        void DamageTaken(Unit* attacker, uint32& damage) override
+        {
+            if (damage >= me->GetHealth())
+            {
+                damage = me->GetHealth() - 1;
+                _events.Reset();
+                me->RemoveAllAuras();
+                me->setFaction(35);
+                me->SetUnitFlags(UnitFlags(UNIT_FLAG_UNK_15 | UNIT_FLAG_IMMUNE_TO_PC));
+                me->AttackStop();
+                attacker->AttackStop();
+                me->_ExitVehicle();
+                attacker->ToPlayer()->KilledMonsterCredit(NPC_MONK_ON_POLE_1);
+                _events.ScheduleEvent(EVENT_DESPAWN, 1000);
+            }
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            _events.Update(diff);
+
+            while (uint32 eventId = _events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_THROW_ROCK:
+                        if (!me->IsWithinMeleeRange(me->GetVictim()))
+                            DoCastVictim(SPELL_THROW_ROCK);
+                        _events.ScheduleEvent(EVENT_THROW_ROCK, 2500);
+                        break;
+                    case EVENT_SWITCH_POLE:
+                        if (!me->IsInCombat())
+                        {
+                            SwitchPole();
+                            _events.ScheduleEvent(EVENT_SWITCH_POLE, urand(15000, 30000));
+                        }
+                        break;
+                    case EVENT_DESPAWN:
+                        // Transform is casted only when in frog pool
+                        if (me->FindNearestCreature(NPC_CURSED_POOL_CONTROLLER, 71.0f, true))
+                            DoCastSelf(SPELL_CURSE_OF_THE_FROG, true);
+                        ClearThreadList();
+                        me->SetWalk(true);
+                        MoveForward(10.0f);
+                        me->DespawnOrUnsummon(3000);
+                        break;
+                }
+            }
+
+            DoMeleeAttackIfReady();
+        }
+
+    private:
+        EventMap _events;
+
+        void SwitchPole()
+        {
+            std::list<Creature*> polesList;
+            std::list<Creature*> polesList2;
+            // This stores objects that are too far away due to big combat reach
+            me->GetCreatureListWithEntryInGrid(polesList, NPC_BALANCE_POLE_1, 1.0f);
+            me->GetCreatureListWithEntryInGrid(polesList2, NPC_BALANCE_POLE_2, 1.0f);
+            // Join both lists with possible different NPC entries
+            polesList.splice(polesList.end(), polesList2);
+            // Convert list to vector, so we can access iterator to be able to shuffle the list
+            std::vector<Creature*> balancePolesList{ std::make_move_iterator(std::begin(polesList)), std::make_move_iterator(std::end(polesList)) };
```

Comment:
You can just randomly advance `std::list::begin` (nitpick)

Reply:
I am not sure I understand.
Advancing the iterator in list in not an option, because the probability of picking random balance pole NPC is not guaranteed than when shuffling the whole list (vector). It's hard to explain why but I'll try.

Imagine you want to find surrounding balance pole NPCs around another NPC within range 1y. So you want to end up with a list of NPCs around another NPC like this:
![image](https://user-images.githubusercontent.com/32191/63434577-a99d4080-c425-11e9-9200-dba923da50bd.png)
You expect to have only those 4 balance pole NPCs, but nope - you end up with having a lot more balance pole NPCs because they have very big combat reach in creature_model_info. So you have something like this:
![image](https://user-images.githubusercontent.com/32191/63436101-61335200-c428-11e9-9f9b-3643d5d0e8bd.png)
Advancing the iterator would do this:
- advance by 2
![image](https://user-images.githubusercontent.com/32191/63436488-11a15600-c429-11e9-9ac9-35ed78f573f0.png)
- advance by 5
![image](https://user-images.githubusercontent.com/32191/63436578-3b5a7d00-c429-11e9-9c6d-2e2774db9eea.png)
will both result in picking this:
![image](https://user-images.githubusercontent.com/32191/63436619-4a412f80-c429-11e9-8c16-f9c364d3c315.png)
because the balance poles are filtered by their spawn position coordinate offsets in the next step and only the first relevant (closest NPC) is picked. So everyting between 0 and 5 will be the same.

But when I shuffle the whole list (vector), I have a guaranty that it will be always random.

Even if this was mega long ago, I do remember that I was asking on the IRC for advice, because I didn't know how to shuffle that list of NPCs, and @Shauren told me to convert it to vector instead and use method RandomShuffle from TC library.